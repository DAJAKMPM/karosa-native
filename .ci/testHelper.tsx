import React, { JSXElementConstructor } from "react";
import { View } from "react-native";
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from "@testing-library/react-native";
import { Formik, FormikValues } from "formik";

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  _themeProp: unknown = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };

  const renderApi = render(children, options);
  const wrapper = renderApi.queryByTestId(wrapperTestID || "wrapper");
  return { wrapper, ...renderApi };
};

export const renderWithFormikWrapper = <T extends FormikValues>(
  initialValues: T,
  children: React.ReactElement<any, string | JSXElementConstructor<any>>
) => {
  return renderWithWrapper(
    <Formik
      initialValues={initialValues}
      onSubmit={() => console.log("submit")}
      component={() => children}
    />
  );
};
