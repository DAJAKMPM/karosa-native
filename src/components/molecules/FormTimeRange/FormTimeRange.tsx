/**
 *
 * FormTimeRange
 * @format
 *
 */

import React, { FC, useState, useCallback } from "react";
import { useField } from "formik";
import { isNil } from "lodash";
import { useUpdateEffect } from "@app/hooks";
import TimeRange from "@app/atoms/TimeRange";

import type { PropsType } from "./types";

const FormTimeRange: FC<PropsType> = (props) => {
  const { startName, endName } = props;

  const [, startMeta, startHelpers] = useField(startName);
  const [, endMeta, endHelpers] = useField(endName);

  const [currentStartValue, setCurrentStartValue] = useState(
    isNil(startMeta.value) ? startMeta.initialValue : startMeta.value
  );

  const [currentEndValue, setCurrentEndValue] = useState(
    isNil(endMeta.value) ? endMeta.initialValue : endMeta.value
  );

  useUpdateEffect(() => {
    setCurrentStartValue(startMeta.value);
  }, [startMeta.value]);

  useUpdateEffect(() => {
    setCurrentEndValue(endMeta.value);
  }, [endMeta.value]);

  const handleStartTimeChange = useCallback(
    (text: string) => {
      setCurrentStartValue(text);
      startHelpers.setValue(text);
      startHelpers.setTouched(true);
    },
    [startHelpers]
  );

  const handleEndTimeChange = useCallback(
    (text: string) => {
      setCurrentEndValue(text);
      endHelpers.setValue(text);
      endHelpers.setTouched(true);
    },
    [endHelpers]
  );

  return (
    <TimeRange
      {...props}
      startValue={currentStartValue}
      endValue={currentEndValue}
      onStartTimeChange={handleStartTimeChange}
      onEndTimeChange={handleEndTimeChange}
    />
  );
};

export default FormTimeRange;
