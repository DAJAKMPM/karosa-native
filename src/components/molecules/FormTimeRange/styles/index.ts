/**
 *
 * @format
 *
 */

import { StyleSheet } from "react-native";

const FormTimeRangeStyles = StyleSheet.create({});

export default FormTimeRangeStyles;
