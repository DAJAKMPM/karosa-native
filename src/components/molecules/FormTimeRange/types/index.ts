/**
 *
 * @format
 *
 */

import { PropsType as TimeRangeType } from "@app/atoms/TimeRange/types";

type Keys = "onStartTimeChange" | "startValue" | "onEndTimeChange" | "endValue";

export type PropsType = Omit<TimeRangeType, Keys> & {
  startName: string;
  endName: string;
};
