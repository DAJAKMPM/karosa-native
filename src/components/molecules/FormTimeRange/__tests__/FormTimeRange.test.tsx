import React from "react";
import { renderWithFormikWrapper } from "@app/ci/testHelper";
import FormTimeRange from "../FormTimeRange";

describe("FormTimeRange Component", () => {
  it("should match snapshot", () => {
    const { toJSON } = renderWithFormikWrapper(
      { startTest: "", endTest: "" },
      <FormTimeRange startName="startTest" endName="endTest" />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
