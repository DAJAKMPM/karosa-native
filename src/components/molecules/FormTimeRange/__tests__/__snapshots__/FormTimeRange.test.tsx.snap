// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTimeRange Component should match snapshot 1`] = `
<View
  testID="wrapper"
>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      keyboardType="numeric"
      onChangeText={[Function]}
      options={
        Object {
          "format": "HH:mm",
        }
      }
      placeholder="HH:mm"
      placeholderTextColor="#0AA351"
      rejectResponderTermination={true}
      returnKeyType="done"
      type="datetime"
      underlineColorAndroid="transparent"
      value=""
    />
    <Text
      style={
        Object {
          "color": "#1C1C1C",
          "fontFamily": "SFProText-Regular",
          "fontSize": 16,
          "fontWeight": "400",
          "lineHeight": 18.75,
          "marginHorizontal": 8,
        }
      }
    >
      -
    </Text>
    <TextInput
      allowFontScaling={true}
      keyboardType="numeric"
      onChangeText={[Function]}
      options={
        Object {
          "format": "HH:mm",
        }
      }
      placeholder="HH:mm"
      placeholderTextColor="#0AA351"
      rejectResponderTermination={true}
      returnKeyType="done"
      type="datetime"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;
